delta-lake/src/main/java/org/apache/iceberg/delta/BaseMigrateDeltaLakeTableAction.java

```diff
@@ -0,0 +1,356 @@
+    // of file(s) were removed
+    // 3. OverwriteFiles - when there are a mix of AddFile and RemoveFile (a DELETE/UPDATE)
+
+    // Create a map of Delta Lake Action (AddFile, RemoveFile, etc.) --> List<Action>
+    Map<String, List<Action>> deltaLakeActionMap =
+        actions.stream()
+            .filter(action -> action instanceof AddFile || action instanceof RemoveFile)
+            .collect(Collectors.groupingBy(a -> a.getClass().getSimpleName()));
+
+    List<DataFile> filesToAdd = Lists.newArrayList();
+    List<DataFile> filesToRemove = Lists.newArrayList();
+    for (Action action : Iterables.concat(deltaLakeActionMap.values())) {
+      DataFile dataFile = buildDataFileFromAction(action, transaction.table());
+      if (action instanceof AddFile) {
+        filesToAdd.add(dataFile);
+      } else if (action instanceof RemoveFile) {
+        filesToRemove.add(dataFile);
+      } else {
+        throw new ValidationException(
+            "The action %s's is unsupported", action.getClass().getSimpleName());
+      }
+    }
+
+    if (filesToAdd.size() > 0 && filesToRemove.size() > 0) {
+      // Overwrite_Files case
+      OverwriteFiles overwriteFiles = transaction.newOverwrite();
+      filesToAdd.forEach(overwriteFiles::addFile);
+      filesToRemove.forEach(overwriteFiles::deleteFile);
+      overwriteFiles.commit();
+    } else if (filesToAdd.size() > 0) {
+      // Append_Files case
+      AppendFiles appendFiles = transaction.newAppend();
+      filesToAdd.forEach(appendFiles::appendFile);
+      appendFiles.commit();
+    } else if (filesToRemove.size() > 0) {
+      // Delete_Files case
+      DeleteFiles deleteFiles = transaction.newDelete();
+      filesToRemove.forEach(deleteFiles::deleteFile);
+      deleteFiles.commit();
+    }
+  }
+
+  public DataFile buildDataFileFromAction(Action action, Table table) {
+    PartitionSpec spec = table.spec();
+    String path;
+    Map<String, String> partitionValues;
+
+    if (action instanceof AddFile) {
+      AddFile addFile = (AddFile) action;
+      path = addFile.getPath();
+      partitionValues = addFile.getPartitionValues();
+    } else if (action instanceof RemoveFile) {
+      RemoveFile removeFile = (RemoveFile) action;
+      path = removeFile.getPath();
+      partitionValues = removeFile.getPartitionValues();
+    } else {
+      throw new IllegalStateException(
+          String.format(
+              "Unexpected action type for Delta Lake: %s", action.getClass().getSimpleName()));
+    }
+
+    String fullFilePath = getFullFilePath(path, deltaLog.getPath().toString());
+
+    if (partitionValues == null) {
+      // For unpartitioned table, the partitionValues should be an empty map rather than null
+      throw new ValidationException("File %s does not specify a partitionValues", fullFilePath);
+    }
+
+    FileFormat format = determineFileFormatFromPath(fullFilePath);
+    FileIO io = table.io();
```

Comment:
I was suggesting a flexible FileIO input for this, but after thinking a bit more about this part of the logic related to FileIO, I feel maybe we should not use the table's IO to read this. 

It will work probably most of the time, but it's logically inconsistent. The reader and writers are different here, the reader is reading the files written in Delta format, and the writer is writing to Iceberg. They could be different. 

The reader is determined by the Hadoop configuration, and the writer is determined by Iceberg catalog configuration. So I think what logically makes more sense is to initialize a HadoopFileIO in the constructor, and directly use it here. 

What do you think?

Reply:
I think its a good idea. Thanks for your suggestion. Also, it is good that by initializing HadoopFileIO in the constructor, we can replace the field hadoopConfiguration by HadoopFileIO, which is more representative of its role to read metrics from data files